direct/data/datasets.py

```diff
@@ -395,6 +400,259 @@ def __broadcast_mask(self, kspace_shape, mask):
+
+            self.data += [(filename, slc) for slc in range(num_slices)]
+
+            self.volume_indices[filename] = range(current_slice_number, current_slice_number + num_slices)
+
+            current_slice_number += num_slices
+
+    @staticmethod
+    def verify_extra_mat_integrity(filename: pathlib.Path, extra_mats: tuple[str]) -> None:
+        if not extra_mats:
+            return
+
+        for key in extra_mats:
+            mat_key, path = extra_mats[key]
+            extra_fn = path / filename.name
+            with h5py.File(extra_fn, "r") as file:
+                _ = file[mat_key].shape
+            return
+
+    def __len__(self) -> int:
+        return len(self.data)
+
+    def get_slice_data(self, filename: PathOrString, slice_no: int, key: str, extra_keys=None):
+        data = h5py.File(filename, "r")
+        shape = data[key].shape
+
+        if self.kspace_context is None:
+            inds = {(i): (k, l) for i, (k, l) in enumerate([(k, l) for k in range(shape[0]) for l in range(shape[1])])}
+            ind = inds[slice_no]
+            curr_data = np.array(data[key][ind[0]][ind[1]])
+        elif self.kspace_context == "slice":
+            # Slice dimension
+            curr_data = np.array(data[key][slice_no])
+        else:
+            # Time dimension
+            curr_data = np.array(data[key][:, slice_no])
+
+        extra_data = {}
+
+        if extra_keys:
+            for extra_key in self.extra_keys:
+                extra_data[extra_key] = data[extra_key][()]
+        data.close()
+        return curr_data, extra_data
+
+    def get_num_slices(self, filename: PathOrString) -> int:
+        num_slices = self.volume_indices[filename].stop - self.volume_indices[filename].start
+        return num_slices
+
+    def __getitem__(self, idx: int) -> dict[str, Any]:  # pylint: disable=too-many-locals
+        filename, slice_no = self.data[idx]
+        filename = pathlib.Path(filename)
+
+        kspace, extra_data = self.get_slice_data(filename, slice_no, key=self.kspace_key, extra_keys=self.extra_keys)
+
+        kspace = kspace["real"] + 1j * kspace["imag"]
+        kspace = np.swapaxes(kspace, -1, -2)
+
+        if kspace.ndim == 2:  # Single-coil data.
+            kspace = kspace[np.newaxis, ...]
+
+        sample = {"kspace": kspace, "filename": str(filename), "slice_no": slice_no}
+
+        if self.compute_mask or (any("mask" in key for key in extra_data)):
+            nx, ny = kspace.shape[-2:]
+            if self.compute_mask:
+                sampling_mask = np.abs(kspace).sum(tuple(range(len(kspace.shape) - 2))) != 0
+
+            else:
+                mask_keys = [key for key in extra_data if "mask" in key]
+                # This will load up randomly a mask if more than one keys
+                mask_key = np.random.choice(mask_keys)
+                sampling_mask = np.array(extra_data[mask_key]).astype(bool)
+                sampling_mask = np.swapaxes(sampling_mask, -1, -2)
```

Comment:
I'm not a fan to have randomness in the dataset itself, that should be part of a transform of some sorts.

Reply:
Removed.